docs/design.md

```diff
@@ -0,0 +1,481 @@
+and one could alternatively use the constant `1 / (id_1 - id_0)` between the
+opcode ID and the first opcode ID to compute the Boolean flag.
+
+Intermediate values used below are not columns but simply variables computed
+from the columns. They don't cost anything.
+
+Constraints are described as arithmetic formulas that should equal 0. The `= 0`
+is omitted for the sake of simplicity.
+
+#### CallAbsImm, Ret
+
+Columns:
+
+- pc
+- fp
+- clock
+- inst_prev_clock
+- opcode_id
+- off0
+- imm
+- op0_prev_clock
+- op0_prev_val
+- op0_plus_one_prev_clock
+- op0_plus_one_prev_val
+
+Intermediate columns:
+
+- `is_ret = opcode_id - CALL_ABS_IMM_ID`
+- `pc_next = imm * (1 - is_ret) + op0_plus_one_prev_val * is_ret`
+- `fp_next = (fp + off0 + 2) * (1 - is_ret) + op0_prev_val * is_ret`
+- `op0_val = fp * (1 - is_ret) + op0_prev_val * is_ret`
+- `op0_plus_one_val = pc * (1 - is_ret) + op0_plus_one_prev_val * is_ret`
+
+Constraints:
+
+- `is_ret * (1 - is_ret)`
+
+Lookups
+
+- update registers
+  - `-Registers(pc, fp)`
+  - `+Registers(pc_next, fp_next)`
+- read instruction from read-only memory
+  - `-Memory(pc, 0, opcode_id, off0, off1)`
+- read/write operands from memory
+  - `-Memory(fp + off0, prev_clock, op0_prev_val)`
+  - `+Memory(fp + off0, clock, op0_val)`
+  - `-Memory(fp + off0 + 1, prev_clock, op0_plus_one_prev_val)`
+  - `+Memory(fp + off0 + 1, clock, op0_plus_one_val)`
+- range check clock difference
+  - `+RangeCheck20(clock - inst_prev_clock - 1)`
+  - `+RangeCheck20(clock - op0_prev_clock - 1)`
+  - `+RangeCheck20(clock - op0_plus_one_prev_clock - 1)`
+
+#### JmpRelImm, JnzFpImm
+
+- registers: pc | fp
+- global: clock
+- instruction: opcode_id | off0 | imm
+- operands: memory[fp + off0]: prev_clock | prev_val
+- memory[fp + off0 + 1]: prev_clock | prev_val
+
+#### StoreAdd, StoreSub, StoreMul, StoreDiv
+
+- registers: pc | fp
+- global: clock
+- instruction: opcode_id | off0 | imm
+- operands: memory[fp + off0]: prev_clock | prev_val
+- memory[fp + off0 + 1]: prev_clock | prev_val
+
+#### StoreDoubleDerefFpFp
+
+- registers: pc | fp
+- global: clock
+- instruction: opcode_id | off0 | imm
+- operands: memory[fp + off0]: prev_clock | prev_val
+- memory[fp + off0 + 1]: prev_clock | prev_val
+
+#### StoreDoubleDerefFp
+
+- registers: pc | fp
+- global: clock
+- instruction: opcode_id | off0 | imm
+- operands: memory[fp + off0]: prev_clock | prev_val
+- memory[fp + off0 + 1]: prev_clock | prev_val
+
+#### StoreFramePointer
+
+- registers: pc | fp
+- global: clock
+- instruction: opcode_id | off0 | imm
+- operands: memory[fp + off0]: prev_clock | prev_val
+- memory[fp + off0 + 1]: prev_clock | prev_val
+
+#### StoreImm
+
+- registers: pc | fp
+- global: clock
+- instruction: opcode_id | off0 | imm
+- operands: memory[fp + off0]: prev_clock | prev_val
+- memory[fp + off0 + 1]: prev_clock | prev_val
+
+### Lookups
+
+This entire paper is drafted with Stwo's constraint framework in mind,
```

Comment:
Should be put in first line of the document with what concept to know from stwo imo